src/connector/src/sink/snowflake.rs

```diff
@@ -0,0 +1,332 @@
+    pub async fn new(
+        config: SnowflakeConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+    ) -> Self {
+        let http_client = SnowflakeHttpClient::new(
+            config.common.account_identifier.clone(),
+            config.common.user.clone(),
+            config.common.database.clone(),
+            config.common.schema.clone(),
+            config.common.pipe.clone(),
+            config.common.rsa_public_key_fp.clone(),
+            config.common.private_key.clone(),
+            HashMap::new(),
+            config.common.s3_path.clone(),
+        );
+
+        let s3_client = SnowflakeS3Client::new(
+            config.common.s3_bucket.clone(),
+            config.common.s3_path.clone(),
+            config.common.aws_access_key_id.clone(),
+            config.common.aws_secret_access_key.clone(),
+            config.common.aws_region.clone(),
+        )
+        .await;
+
+        let max_batch_row_num = config
+            .common
+            .max_batch_row_num
+            .clone()
+            .parse::<u32>()
+            .expect("failed to parse `snowflake.max_batch_row_num` as a `u32`");
+
+        Self {
+            config,
+            schema: schema.clone(),
+            pk_indices,
+            is_append_only,
+            http_client,
+            s3_client,
+            row_encoder: JsonEncoder::new(
+                schema,
+                None,
+                super::encoder::DateHandlingMode::String,
+                TimestampHandlingMode::String,
+                TimestamptzHandlingMode::UtcString,
+                TimeHandlingMode::String,
+            ),
+            row_counter: 0,
+            payload: String::new(),
+            max_batch_row_num,
+            epoch: 0,
+            // Start from 0, i.e., `RW_SNOWFLAKE_S3_SINK_FILE_0`
+            sink_file_suffix: 0,
+        }
+    }
+
+    /// Note that we shall NOT reset the `sink_file_suffix`
+    /// since we need to incrementally keep the sink
+    /// file *unique*, otherwise snowflake will not
+    /// sink it from external stage (i.e., s3)
+    fn reset(&mut self) {
+        self.payload.clear();
+        self.row_counter = 0;
+    }
+
+    fn at_sink_threshold(&self) -> bool {
+        self.row_counter >= self.max_batch_row_num
+    }
+
+    fn append_only(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            assert_eq!(op, Op::Insert, "expect all `op(s)` to be `Op::Insert`");
+            let row_json_string = Value::Object(self.row_encoder.encode(row)?).to_string();
+            self.payload.push_str(&row_json_string);
+            self.row_counter += 1;
+        }
+        Ok(())
+    }
+
+    fn update_epoch(&mut self, epoch: u64) {
+        self.epoch = epoch;
+    }
+
+    /// Construct the *unique* file suffix for the sink
+    fn file_suffix(&self) -> String {
+        format!("{}_{}", self.epoch, self.sink_file_suffix)
```

Comment:
The `sink_file_suffix` is local to each parallelism of snowflake writer, so there might be a same sink_file_suffix across multiple parallelisms. Do we need to ensure uniqueness across multiple parallelisms?

Reply:
as long as `epoch` remains different, everthing should be fine - actually I'm thinking to only include the `epoch` as the unique identifier for the s3 intermediate file(s), the `sink_file_suffix` does not really matter, cc @xxhZs @fuyufjh.

refer: https://github.com/risingwavelabs/risingwave/pull/15429#discussion_r1522117544

Reply:
But different parallelisms share the same epochs. Let's say we are now at epoch 233, and we have two sink parallelisms. The first file name in this epoch of both executors are both `<s3_path>/233-0`. In this case, will the later written one overwrite the firstly written one?

Reply:
Overwritten will not happen, but it's even worse 🫠 - the latter one upload the data to the same file on external S3, and we trigger the `insertFiles` request to snowflake pipe; but the file name remains the same, snowflake then implicitly treats this as redundant, and the new version will not be loaded into the pipe.

Thus, we need another identifier for the parallel writer(s) scenario - any suggestion?

Reply:
[uuid](https://docs.rs/uuid/latest/uuid/) is fine. To help debug we can use `<epoch>-<uuid>` as the file name.